File: Python/289. Game of Life.py
class Solution:
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        # if board[i][j]==0 and its value should be changed, set board[i][j]=2;
        # if board[i][j]==1 and its value should be changed, set board[i][j]=3;

        row = len(board)
        col = len(board[0])
        for i in range(row):
            for j in range(col):
                n = 0
                if i-1 >= 0:
                    n += board[i-1][j] % 2
                    if j-1 >= 0:
                        n += board[i-1][j-1] % 2
                    if j+1 < col:
                        n += board[i-1][j+1] % 2
                if i+1 < row:
                    n += board[i+1][j] % 2
                    if j-1 >= 0:
                        n += board[i+1][j-1] % 2
                    if j+1 < col:
                        n += board[i+1][j+1] % 2
                if j-1 >= 0:
                    n += board[i][j-1] % 2
                if j+1 < col:
                    n += board[i][j+1] % 2

                if board[i][j] == 1:
                    if n < 2 or n > 3:
                        board[i][j] = 3
                elif board[i][j] == 0:
                    if n == 3:
                        board[i][j] = 2

        # update
        for i in range(row):
            for j in range(col):
                if board[i][j] == 2:
                    board[i][j] = 1
                elif board[i][j] == 3:
                    board[i][j] = 0

File: 算法/C++/289. Game of Life.cpp
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        // if board[i][j]==0 and its value should be changed, set board[i][j]=2;
        // if board[i][j]==1 and its value should be changed, set board[i][j]=3;
        int m = board.size();
        int n = board[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int cnt = 0;
                if (i - 1 >= 0) {
                    cnt += board[i - 1][j] & 1;
                    if (j - 1 >= 0)
                        cnt += board[i - 1][j - 1] & 1;
                    if (j + 1 < n)
                        cnt += board[i - 1][j + 1] & 1;

                }
                if (i + 1 < m) {
                    cnt += board[i + 1][j] & 1;
                    if (j - 1 >= 0)
                        cnt += board[i + 1][j - 1] & 1;
                    if (j + 1 < n)
                        cnt += board[i + 1][j + 1] & 1;
                }
                if (j - 1 >= 0)
                    cnt += board[i][j - 1] & 1;
                if (j + 1 < n)
                    cnt += board[i][j + 1] & 1;

                if (board[i][j] == 1) {
                    if (cnt < 2 || cnt > 3)
                        board[i][j] = 3;
                }
                else if (board[i][j] == 0) {
                    if (cnt == 3)
                        board[i][j] = 2;
                }

            }
        }
        //update
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 2)
                    board[i][j] = 1;
                else if (board[i][j] == 3)
                    board[i][j] = 0;

            }
        }


    }
};